eth/handler.go

```diff
@@ -413,13 +436,21 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 			}
 		}
 		if len(headers) == 1 {
-			if err := pm.chainConfig.HeaderCheck(headers[0]); err != nil {
+			header := headers[0]
+			err := pm.chainConfig.HeaderCheck(header)
+			if err != nil {
 				pm.removePeer(p.id)
 				return err
 			}
+			if !header.Hash().IsEmpty() {
+				p.MarkHeader(header.Hash())
```

Comment:
@whilei Please be careful about this. This is a potential memory leak and could lead to DoS attack.

Consider a malicious client that constantly send valid headers using the BlockHeaders message of length 1. Then this knownHeaders set will be filled all the way up and use a lot of memory.